### ISSUE TITLE
Error that Tuples of length greater than one are incompatible with "__add__"

### ISSUE THREAD

**Issue by notatallshaw** [microsoft/pyright#4613]
**Describe the bug**

If you make a Protocol for `__add__` then Pyright complains that Tuples of length > 1 are not compatible.

**To Reproduce**

1. Create a Python file like this:

```python
from typing import Protocol, TypeVar, Any

T = TypeVar("T")

class Addable(Protocol):
    def __add__(self: T, other: T, /) -> T:
        ...

def add(a: Addable, b: Addable) -> Any:
    return a + b

add((1,), (2,))
add((1, 2), (3, 4))
```

2. Run `pyright` on this file

3. Get an error like:

```
addtest.py
  addtest.py:16:12 - error: Argument of type "tuple[Literal[1], Literal[2]]" cannot be assigned to parameter "a" of type "Addable" in function "add"
    "tuple[Literal[1], Literal[2]]" is incompatible with protocol "Addable"
      "__add__" is an incompatible type
        No overloaded function matches type "(other: tuple[Literal[1], Literal[2]], /) -> tuple[Literal[1], Literal[2]]" (reportGeneralTypeIssues)
  addtest.py:16:20 - error: Argument of type "tuple[Literal[3], Literal[4]]" cannot be assigned to parameter "b" of type "Addable" in function "add"
    "tuple[Literal[3], Literal[4]]" is incompatible with protocol "Addable"
      "__add__" is an incompatible type
        No overloaded function matches type "(other: tuple[Literal[3], Literal[4]], /) -> tuple[Literal[3], Literal[4]]" (reportGeneralTypeIssues)
2 errors, 0 warnings, 0 informations
```

**Expected behavior**

No errors, like `mypy --strict`:

```
Success: no issues found in 1 source file
```

**VS Code extension or command-line**

Both, using the latest released versions of everything as of today.

**Context**

I appreciate the code sample itself doesn't make much sense. I'm just experimenting with how one would go about Type Hinting functions that rely heavily on Duck Typing.

**Comment by erictraut** [comment]
I think pyright's behavior is correct.

Let's simplify the code.

```python
x1: Addable = (1,) # OK
x2: Addable = (1, 2) # Error
```

To understand what's going on, we need to look at the definition of `tuple.__add__`. It's an overloaded method with two overloads:

```python
class tuple(Sequence[_T_co]):
    ...
    @overload
    def __add__(self, __x: tuple[_T_co, ...]) -> tuple[_T_co, ...]: ...
    @overload
    def __add__(self, __x: tuple[_T, ...]) -> tuple[_T_co | _T, ...]: ...
```

The type of `(1, 2)` is `tuple[Literal[1], Literal[2]]`. In this case, the type variable `_T_co` takes on the value of `Literal[1, 2]`. That means:
* The specialized version of `tuple.__add__` is `(_: tuple[Literal[1, 2], ...], /) -> tuple[Literal[1, 2], ...]`
* The specialized version of `Addable.__add__` is `(_: tuple[Literal[1], Literal[2]], /) -> tuple[Literal[1], Literal[2]]`

The specialized `Addable.__add__` is not type compatible with the specialized `tuple.__add__` because `tuple[Literal[1, 2], ...]` is not assignable to `tuple[Literal[1], Literal[2]]`.

This problem doesn't occur if all elements in the tuple are the same type (e.g. `(1, )` or `(1, 1)`).

```python
x3: Addable = (1, 1) # OK

def func(a: tuple[int, int], b: tuple[int, str]):
    x4: Addable = a # OK
    x5: Addable = b # Error
```

I don't know why mypy doesn't consider this an error.

**Comment by notatallshaw** [comment]
Thanks for the detailed explanation, it's a little subtle but I think I follow, I will raise an issue with MyPy.

**Comment by erictraut** [comment]
It's possible that mypy has some special-case logic that overrides the type definition for `tuple.__add__` in `builtins.pyi`. We generally don't do that in pyright because it leads to fragility and maintainability issues.

**Comment by sad2project** [comment]
So, does this mean that if you type hinted the tuples as ints instead of literals, it should work? i.e.:

    a: tuple[int, int] = (1, 2)
    b: tuple[int, int] = (3, 4)

    add((1, 2), (3, 4))

**Comment by notatallshaw** [comment]
Im not at a computer right now so I can't test but the resolution of the MyPy issue that MyPy implicitly types the tuples as `tuple[int, ...]` and therefore doesn't throw an error. I assume I'd you explicitly type the tuples as that then Pyright should pass.

I've not dug in to why MyPy and Pyright have different implicit type handling or if there is a standard on implicit typing.

**Comment by erictraut** [comment]
Refer to [this documentation](https://github.com/microsoft/pyright/blob/main/docs/mypy-comparison.md#inference-of-tuple-expressions) for details.

**Comment by kcdodd** [comment]
At the risk of looking really dumb here, I was trying to understand this logic but I couldn't reproduce the simplified examples as even the single-element tuple causes me an issue. I then tried to understand how "tuple" was defined as explained above, but that also doesn't seem assignable. A plain "Sequence" is ok, but not the subclass. Besides that, the "tuple" typeshed code seems invalid if "tuple" is replaced with the name of the class, since the base Generic type parameters don't actually match how its used with the variadic ",,,".  

```python
from __future__ import annotations
from typing import TypeVar, Protocol
from collections.abc import Sequence

T = TypeVar("T")

class Addable(Protocol):
    def __add__(self: T, other: T, /) -> T:
        ...

x1: Addable = (1,) # ERROR
x2: Addable = (1, 2) # ERROR

_T_co = TypeVar("_T_co", covariant=True)

class MySequence(Sequence[_T_co]):
  ...

y1: Sequence[int] = (1, 2) # OK
y2: MySequence[int] = (1,) # ERROR
y3: MySequence[int] = (1, 2) # ERROR
```

```
/home/cdodd/proj/test.py
  /home/cdodd/proj/test.py:11:15 - error: Expression of type "tuple[Literal[1]]" is incompatible with declared type "Addable"
    "tuple[Literal[1]]" is incompatible with protocol "Addable"
      "__add__" is an incompatible type
        No overloaded function matches type "(other: tuple[Literal[1]], /) -> tuple[Literal[1]]" (reportAssignmentType)
  /home/cdodd/proj/test.py:12:15 - error: Expression of type "tuple[Literal[1], Literal[2]]" is incompatible with declared type "Addable"
    "tuple[Literal[1], Literal[2]]" is incompatible with protocol "Addable"
      "__add__" is an incompatible type
        No overloaded function matches type "(other: tuple[Literal[1], Literal[2]], /) -> tuple[Literal[1], Literal[2]]" (reportAssignmentType)
  /home/cdodd/proj/test.py:20:23 - error: Expression of type "tuple[Literal[1]]" is incompatible with declared type "MySequence[int]"
    "tuple[Literal[1]]" is incompatible with "MySequence[int]" (reportAssignmentType)
  /home/cdodd/proj/test.py:21:23 - error: Expression of type "tuple[Literal[1], Literal[2]]" is incompatible with declared type "MySequence[int]"
    "tuple[Literal[1], Literal[2]]" is incompatible with "MySequence[int]" (reportAssignmentType)
4 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
`tuple` is a [nominal](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-nominal) type, as opposed to a [structural](https://typing.readthedocs.io/en/latest/spec/glossary.html#term-structural) type. It is a subclass of `Collection` which is a subclass of `Sequence`.

`MySequence` is also a subclass of `Sequence`, but it is not a subtype of `tuple`, nor is `tuple` a subtype of `MySequence`, so there's no type compatibility between the two.

If you're interested in learning more about type concepts that underlie the Python type system, you might want to read [this section](https://typing.readthedocs.io/en/latest/spec/concepts.html) of the Python typing spec.

**Comment by kcdodd** [comment]
Oh right, MySequence isn't like protocol, and isn't a parent. But still, I'm confused by the typeshed tuple generic not being a variadic, yet is used as one, inheriting from Sequence[T] like "MySequence". What I was trying to show is that "MySequence[int, ...]" is not valid, and was not possible to write the same definition for __add__ like for "tuple[int, ...]".

**Comment by kcdodd** [comment]
Maybe a more explicit example would help

```python
from __future__ import annotations
from typing import TypeVar, overload
from collections.abc import Sequence

_T = TypeVar("_T")
_T_co = TypeVar("_T_co", covariant=True)

# OK
class MyTuple1(Sequence[_T_co]):
    @overload
    def __add__(self, value: tuple[_T_co, ...], /) -> tuple[_T_co, ...]: ...
    @overload
    def __add__(self, value: tuple[_T, ...], /) -> tuple[_T_co | _T, ...]: ...

    def __add__(self, value):
      return self + value

# ERROR
class MyTuple2(Sequence[_T_co]):
    @overload
    def __add__(self, value: MyTuple2[_T_co, ...], /) -> MyTuple2[_T_co, ...]: ...
    @overload
    def __add__(self, value: MyTuple2[_T, ...], /) -> MyTuple2[_T_co | _T, ...]: ...

    def __add__(self, value):
      return self + value
```

```
/home/cdodd/proj/test.py
  /home/cdodd/proj/test.py:21:46 - error: Too many type arguments provided for "MyTuple2"; expected 1 but received 2 (reportInvalidTypeArguments)
  /home/cdodd/proj/test.py:21:46 - error: "..." is not allowed in this context
  /home/cdodd/proj/test.py:21:74 - error: Too many type arguments provided for "MyTuple2"; expected 1 but received 2 (reportInvalidTypeArguments)
  /home/cdodd/proj/test.py:21:74 - error: "..." is not allowed in this context
  /home/cdodd/proj/test.py:23:43 - error: Too many type arguments provided for "MyTuple2"; expected 1 but received 2 (reportInvalidTypeArguments)
  /home/cdodd/proj/test.py:23:43 - error: "..." is not allowed in this context
  /home/cdodd/proj/test.py:23:76 - error: Too many type arguments provided for "MyTuple2"; expected 1 but received 2 (reportInvalidTypeArguments)
  /home/cdodd/proj/test.py:23:76 - error: "..." is not allowed in this context
8 errors, 0 warnings, 0 informations 
```

**Comment by erictraut** [comment]
The `tuple` class requires significant special-casing in type checkers. As you've pointed out, it's variadic, so it can't be described in the typeshed stub. There's an entire [chapter](https://typing.readthedocs.io/en/latest/spec/tuples.html) in the typing spec dedicated to `tuple` and its special properties.

I'm not sure what you are trying to do, but perhaps what you're looking for is a [TypeVarTuple](https://typing.readthedocs.io/en/latest/spec/generics.html#typevartuple), the variadic form of a `TypeVar`. If you want additional help, I recommend starting a conversation in the discussions section rather than posting to a closed bug report.

**Comment by kcdodd** [comment]
Understood. The original issue was how to make a Protocol that matches a tuple based on the operator (in addition to other "similar" types).
